[ACS - Chat] GA Attachments to ChatMessage

jimchou-dev opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).
- Introduce "attachments" property and ChatAttachment
- Based on previous approved public preview https://github.com/Azure/azure-rest-api-specs/pull/22331

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: 
* Previous API Spec Doc: 
* Updated paths: 

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/26018

jimchou-dev committed:
Update Swagger for receive image from Teams user

mikekistler reviewed (changes requested):
Please fix the Avocado errors.

mikekistler commented on specification/communication/data-plane/Chat/stable/2023-11-07/communicationserviceschat.json:
```diff
@@ -0,0 +1,1793 @@
+    }
+  },
+  "definitions": {
+    "CommunicationUserIdentifierModel": {
+      "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationUserIdentifierModel"
+    },
+    "CommunicationCloudEnvironment": {
+      "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationCloudEnvironmentModel"
+    },
+    "MicrosoftTeamsUserIdentifierModel": {
+      "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/MicrosoftTeamsUserIdentifierModel"
+    },
+    "CommunicationIdentifierModel": {
+      "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationIdentifierModel"
+    },
```
In 2023-07-01-preview, these refs all used the more recent `Common/stable/2022-07-13` versions of these definitions. Why did you decide to go back to using the older ones?

jimchou-dev replied:
That's a good catch, thank you was copying the folder from previous stable.

jimchou-dev committed:
fix pr comments

jimchou-dev committed:
change default api version

JeffreyRichter commented:
I made some comments in the APIView.

jimchou-dev commented:
> I made some comments in the APIView.
@JeffreyRichter replied to those comments, please let me know if they can be resolved

JeffreyRichter reviewed (approved):
I'm approving the PR but the unknown enum value should be removed if it didn't already GA previously. See comment in APIView.

jimchou-dev committed:
rename contentType to extension

jimchou-dev commented:
Hi @JeffreyRichter, I have updated the API spec with 2 changes:
- Removed the free-form metadata field (just double check it wasn't GA yet)
- Rename "contentType" to "extension" in ChatAttachment

When you have time can you please review and approve of this REST API again.

Also thanks for expediting this review with async feedback, was wondering what do I do with the scheduled meeting at Nov 01, 2023 09:00 AM https://github.com/Azure/azure-rest-api-specs/issues/26018

jimchou-dev committed:
add back metadata, seem like it was GA-ed previously

jpeng-ms committed:
Update communicationserviceschat.json

jimchou-dev committed:
remove file from GA

mikekistler commented on specification/communication/data-plane/Chat/stable/2023-11-07/communicationserviceschat.json:
```diff
@@ -0,0 +1,1791 @@
+          "type": "string",
+          "example": "A27C7079-626F-4D23-8BEE-582F2C376B5A"
+        },
+        "attachmentType": {
+          "$ref": "#/definitions/AttachmentType"
+        },
```
Currently this field can only have one value, "image". What happens when a new type is added? Does this become polymorphic? If so we should describe it as polymorphic from the start.

jimchou-dev replied:
AttachmentType is going to be `image` and `file`, but because we didn't receive customer validation yet for file-sharing feature, so we decide to GA only image for now.
How do we describe to indicate AttachmentType will be polymorphic?

mikekistler replied:
You would specify that the `attachmentType` property is a `discriminator` and then have separate schemas for each attachment type. See the OpenAPI docs here

https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#composition-and-inheritance-polymorphism

jimchou-dev replied:
Thanks for sharing the resource, after having a look, it's actually not going to be polymorphism, both `image` and `file` attachmentType have exactly the same ChatAttachment properties, the only difference is will have a different attachmentType

mikekistler reviewed (commented):
Please clarify the intent of the "attachmentType" field.

mikekistler reviewed (approved):
Looks good. 👍

jimchou-dev committed:
remove uri format

jimchou-dev committed:
Rename enum name to ChatAttachment and revert uri type

jpeng-ms commented on specification/communication/data-plane/Chat/stable/2023-11-07/communicationserviceschat.json:
```diff
@@ -0,0 +1,1791 @@
+        "attachmentType": {
+          "$ref": "#/definitions/ChatAttachmentType"
+        },
+        "extension": {
```
@mikekistler, @JeffreyRichter, we are getting some feedback on naming of this property. I was wondering if "fileType" makes sense?

an example would be:

```
"attachments": [
                    {
                        "id": "efecf033e5198a9c15c92a3163dc90e5",
                        "attachmentType": "image",
                        "fileType": "png",
                        "name": "img",
                        "url": "https://contoso.azurewebsites.net/chat/threads/19:efecf033e5198a9c15c92a3163dc90e5@thread.v2/images/0efecf033e5198a9c15c92a3163dc90e5/views/original?api-version=2023-11-07",
                        "previewUrl": ""https://contoso.azurewebsites.net/chat/threads/19:efecf033e5198a9c15c92a3163dc90e5@thread.v2/images/0efecf033e5198a9c15c92a3163dc90e5/views/small?api-version=2023-11-07"
                    },
                   {
                        "id": "efecf033e5198a9c15c92a3163dc90e5",
                        "attachmentType": "file",
                        "fileType": "pdf",
                        "name": "report.pdf",
 
                        "previewUrl": "[https://contoso.sharepoint.com/:personal/contoso_inc/efecf033e5198a9c15c92a3163dc90e5](https://contoso.sharepoint.com/:personal/contoso_inc/efecf033e5198a9c15c92a3163dc90e5?xsdata=MDV8MDF8fDgzZThkZTIxZjM1ODQxNzlmOTUzMDhkYmUyMzlmMGM3fDcyZjk4OGJmODZmMTQxYWY5MWFiMmQ3Y2QwMTFkYjQ3fDB8MHw2MzgzNTI1MTA2ODM3MDc3MzF8VW5rbm93bnxWR1ZoYlhOVFpXTjFjbWwwZVZObGNuWnBZMlY4ZXlKV0lqb2lNQzR3TGpBd01EQWlMQ0pRSWpvaVYybHVNeklpTENKQlRpSTZJazkwYUdWeUlpd2lWMVFpT2pFeGZRPT18MXxMM1JsWVcxekx6RTVPbVkyWkRVeVlXTTJORFkxWXpRd1pXRTRNR1JqT0RaaU9HSmxNemd5TldGaFFIUm9jbVZoWkM1emEzbHdaUzlqYUdGdWJtVnNjeTh4T1RwbU5tUTFNbUZqTmpRMk5XTTBNR1ZoT0RCa1l6ZzJZamhpWlRNNE1qVmhZVUIwYUhKbFlXUXVjMnQ1Y0dVdmJXVnpjMkZuWlhNdk1UWTVPVFkxTkRJMk5qZ3lOUT09fGY5Njg2Y2JlNGJhZjQ0YmRmOTUzMDhkYmUyMzlmMGM3fDg4NjE4ZjkzNzZkYTQzMjhhNDMzOThmY2E0MWQ4MTgw&sdata=R0Q1Q0pLcXFaRzJYbjQ1bWNOZTVGMXZ6R2U0WWIzN2d1UnJ1aE9wL0pVZz0%3D&ovuser=72f988bf-86f1-41af-91ab-2d7cd011db47%2Cjopeng%40microsoft.com)["](https://global.chat.ppe.communication.microsoft.com/chat/threads/19:meeting_ZDk0ODlhNTgtMzAxNS00MzE5LTg2MGYtYzM4MGJkZmM5Njk5@thread.v2/messages/1677921907914/teamsInterop/images/0-canarycus-d3-efecf033e5198a9c15c92a3163dc90e5/views/small?api-version=2023-04-01-preview%22)
                    },
                   {
                        "id": "efecf033e5198a9c15c92a3163dc90e5",
                        "attachmentType": "file",
 
                        "name": "file_with_no_extension",
                        "previewUrl": "[https://contoso.sharepoint.com/:personal/contoso_inc/efecf033e5198a9c15c92a3163dc90e5](https://contoso.sharepoint.com/:personal/contoso_inc/efecf033e5198a9c15c92a3163dc90e5?xsdata=MDV8MDF8fDgzZThkZTIxZjM1ODQxNzlmOTUzMDhkYmUyMzlmMGM3fDcyZjk4OGJmODZmMTQxYWY5MWFiMmQ3Y2QwMTFkYjQ3fDB8MHw2MzgzNTI1MTA2ODM3MDc3MzF8VW5rbm93bnxWR1ZoYlhOVFpXTjFjbWwwZVZObGNuWnBZMlY4ZXlKV0lqb2lNQzR3TGpBd01EQWlMQ0pRSWpvaVYybHVNeklpTENKQlRpSTZJazkwYUdWeUlpd2lWMVFpT2pFeGZRPT18MXxMM1JsWVcxekx6RTVPbVkyWkRVeVlXTTJORFkxWXpRd1pXRTRNR1JqT0RaaU9HSmxNemd5TldGaFFIUm9jbVZoWkM1emEzbHdaUzlqYUdGdWJtVnNjeTh4T1RwbU5tUTFNbUZqTmpRMk5XTTBNR1ZoT0RCa1l6ZzJZamhpWlRNNE1qVmhZVUIwYUhKbFlXUXVjMnQ1Y0dVdmJXVnpjMkZuWlhNdk1UWTVPVFkxTkRJMk5qZ3lOUT09fGY5Njg2Y2JlNGJhZjQ0YmRmOTUzMDhkYmUyMzlmMGM3fDg4NjE4ZjkzNzZkYTQzMjhhNDMzOThmY2E0MWQ4MTgw&sdata=R0Q1Q0pLcXFaRzJYbjQ1bWNOZTVGMXZ6R2U0WWIzN2d1UnJ1aE9wL0pVZz0%3D&ovuser=72f988bf-86f1-41af-91ab-2d7cd011db47%2Cjopeng%40microsoft.com)["](https://global.chat.ppe.communication.microsoft.com/chat/threads/19:meeting_ZDk0ODlhNTgtMzAxNS00MzE5LTg2MGYtYzM4MGJkZmM5Njk5@thread.v2/messages/1677921907914/teamsInterop/images/0-canarycus-d3-efecf033e5198a9c15c92a3163dc90e5/views/small?api-version=2023-04-01-preview%22)
                    }
]
```

JeffreyRichter replied:
I'm fine with fileType

jpeng-ms committed:
Update communicationserviceschat.json

JoshuaLai committed:
Merge branch 'main' into chat-inline-image-ga

tjprescott commented:
@mikekistler @JeffreyRichter if either (or both) of you have approved on behalf of the Stewardship board, please add the SignedOff label.
